File: android/LocationTracker/src/main/java/ca/mcmaster/capstone/monitoralgorithm/Conjunct.java
package ca.mcmaster.capstone.monitoralgorithm;

import ca.mcmaster.capstone.monitoralgorithm.tree.BooleanExpressionTree;
import ca.mcmaster.capstone.monitoralgorithm.tree.parser.BooleanExpressionParser;
import ca.mcmaster.capstone.networking.structures.NetworkPeerIdentifier;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.ToString;

@EqualsAndHashCode @ToString
public class Conjunct {
    public static enum Evaluation {TRUE, FALSE, NONE}

    @NonNull @Getter private final NetworkPeerIdentifier ownerProcess;
    @NonNull private final BooleanExpressionTree expression;

    public Conjunct(@NonNull final NetworkPeerIdentifier ownerProcess, @NonNull final String expression) {
        this.ownerProcess = ownerProcess;
        this.expression = BooleanExpressionParser.INSTANCE.parse(expression);
    }

    public Evaluation evaluate(@NonNull final ProcessState state) {
        if (!ownerProcess.equals(state.getId())) {
            return Evaluation.NONE;
        }
        return expression.evaluate(state);
    }
}

File: ios/capstone/capstone/Conjunct.swift
//
//  Conjunct.swift
//  capstone
//
//  Created by Darren Kitamura on 2015-01-15.
//  Copyright (c) 2015 DKitamura. All rights reserved.
//

import Foundation


class Conjunct {
    
      enum Evaluation {
        case TRUE
        case FALSE
        case NONE
    }
    
    //Getter
    final private var ownerProcess: NetworkPeerIdentifier
    
    final private var expression: BooleanExpressionTree
    
    init(ownerProcess: NetworkPeerIdentifier, expression: BooleanExpressionTree) {
        self.ownerProcess = ownerProcess
        self.expression = expression
    }
    
    public Evaluation evaluate(state: ProcessState) {
        if(ownerProcess != state.getId()) {
            return Evaluation.NONE
        }
        return expression.evaluate(state)
    }
    
}
